leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/CaliforniumEndpointsManager.java

```diff
@@ -137,7 +139,60 @@ public boolean isTrusted(RawPublicKeyIdentity id) {
+                // - must do PKIX validation with trustStore
+                // - target certificate must match what is provided certificate in server info
+                //
+                // 2: Certificate usage 2 ("trust anchor assertion")
+                // - trustStore is only the provided certificate in server info
+                // - must do PKIX validation with trustStore
+                //
+                // 3: Certificate usage 3 ("domain-issued certificate") (default mode if missing)
+                // - no trustStore used in this mode
+                // - target certificate must match what is provided certificate in server info
+
+                final CertificateUsage certificateUsage = serverInfo.certificateUsage != null ?
+                        serverInfo.certificateUsage :
+                        CertificateUsage.DOMAIN_ISSUER_CERTIFICATE;
+                final Certificate expectedServerCertificate = serverInfo.serverCertificate;
+                X509Certificate[] newTrustedCertificates = null;
+                if (certificateUsage == CertificateUsage.CA_CONSTRAINT) {
+                    // - trustStore is combination of client's configured trust store and provided certificate in server info
+                    ArrayList<X509Certificate> newTrustedCertificatesList = new ArrayList<>();
+                    if (this.trustStore != null) {
+                        newTrustedCertificatesList.addAll(CertPathUtil.toX509CertificatesList(this.trustStore));
+                    }
+                    if (serverInfo.serverCertificate instanceof X509Certificate) {
+                        newTrustedCertificatesList.add((X509Certificate) serverInfo.serverCertificate);
+                    }
+                    newTrustedCertificates = newTrustedCertificatesList.toArray(new X509Certificate[0]);
+                } else if (certificateUsage == CertificateUsage.SERVICE_CERTIFICATE_CONSTRAINT) {
+                    // - trustStore is client's configured trust store
+                    if (this.trustStore != null) {
+                        newTrustedCertificates = CertPathUtil.toX509CertificatesList(this.trustStore).toArray(new X509Certificate[0]);
+                    }
+                } else if (certificateUsage == CertificateUsage.TRUST_ANCHOR_ASSERTION) {
+                    // - trustStore is only the provided certificate in server info
+                    if (serverInfo.serverCertificate instanceof X509Certificate) {
+                        newTrustedCertificates = new X509Certificate[] {
+                                (X509Certificate) serverInfo.serverCertificate };
+                    }
+                }
+
+                final X509Certificate[] trustedCertificates = newTrustedCertificates;
+
+                newBuilder.setCertificateVerifier(
+                        new DefaultLeshanCertificateVerifier(serverInfo.serverCertificate, trustedCertificates,
```

Comment:
With latest changes it is still in one class.

Were you thinking inline class or 4 different classes with one base class?

What I would like to keep is this summary what describes how it works in one place.

Reply:
I was thinking about 4 different class (each class with a copy paste from IETF description).
Each class could have @see which point to other implementation.

My feeling this would be better for separations of concern.

Eventually, those class could be moved in californium project.

> What I would like to keep is this summary what describes how it works in one place.

Eventually if you think it worth it you can keep it where you create instance of `CertificateVerifier` depending on certificate usage ?

Does it sound good to you ?

Reply:
Something like this?